File: 复现的方法/基于阈值的/真实数据集测试/test_real_data.py
"""
基于阈值方法 - 真实数据集测试脚本
对真实数据集进行去噪处理并保存结果
"""
import os
import sys
import scipy.io
import numpy as np
from time import time
import shutil

# 添加路径
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
grandparent_dir = os.path.dirname(os.path.dirname(current_dir))  # 复现的方法目录
sys.path.insert(0, parent_dir)
sys.path.insert(0, current_dir)
sys.path.insert(0, grandparent_dir)  # 添加以访问load_real_dataset_split

from SingleDenoise_CORRECTED import eog_removal_corrected
from real_data_config import *
from load_real_dataset_split import load_real_dataset_split


def load_real_data():
    """加载真实数据（只加载测试集）"""
    # 使用统一的数据划分函数，只返回测试集
    test_data = load_real_dataset_split(
        data_path=REAL_DATA_PATH,
        data_key=DATA_KEY,
        return_train=False  # 只需要测试集
    )
    return test_data


def process_all_samples(contaminated_data, fs=250):
    """
    处理所有样本
    
    Args:
        contaminated_data: [n_samples, n_timepoints]
        fs: 采样率
    
    Returns:
        predictions: 去噪后的信号
        time_per_sample: 单样本平均处理时间
    """
    n_samples, n_timepoints = contaminated_data.shape
    print(f"\n开始处理 {n_samples} 个样本...")
    
    predictions = np.zeros_like(contaminated_data)
    
    start_time = time()
    
    for i in range(n_samples):
        signal = contaminated_data[i, :]
        clean_signal = eog_removal_corrected(signal, fs=fs, visualize=False)
        predictions[i, :] = clean_signal
        
        # 进度显示
        if (i + 1) % 10 == 0:
            elapsed = time() - start_time
            avg_time = elapsed / (i + 1)
            remaining = avg_time * (n_samples - i - 1)
            print(f"  已处理 {i+1}/{n_samples} 样本 "
                  f"(平均 {avg_time:.3f}s/样本, 预计剩余 {remaining/60:.1f}分钟)")
    
    total_time = time() - start_time
    time_per_sample = total_time / n_samples
    
    print(f"\n处理完成!")
    print(f"  总耗时: {total_time:.1f}秒")
    print(f"  单样本平均时间: {time_per_sample:.3f}秒")
    
    return predictions, time_per_sample


def main():
    print("=" * 80)
    print("基于阈值方法 - 真实数据集测试")
    print("=" * 80)
    
    # 加载数据
    test_x = load_real_data()
    
    # 处理所有样本
    predictions, time_per_sample = process_all_samples(test_x, fs=SAMPLING_RATE)
    
    print(f'\n测试完成！')
    print(f'  预测结果形状: {predictions.shape}')
    print(f'  平均处理时间: {time_per_sample*1000:.2f} ms/样本')
    
    # 计算伪影（原始信号 - 去噪信号）
    artifacts = test_x - predictions
    
    # 验证解耦一致性
    print('\n验证解耦一致性...')
    reconstructed = predictions + artifacts
    consistency_error = np.mean((reconstructed - test_x) ** 2)
    print(f'  重建一致性 MSE: {consistency_error:.6f}')
    
    # 统计信息
    print('\n统计信息:')
    original_std = np.std(test_x)
    cleaned_std = np.std(predictions)
    artifact_std = np.std(artifacts)
    print(f'  原始信号标准差: {original_std:.4f}')
    print(f'  去噪信号标准差: {cleaned_std:.4f}')
    print(f'  伪影标准差: {artifact_std:.4f}')
    
    power_reduction = (np.mean(test_x ** 2) - np.mean(predictions ** 2)) / np.mean(test_x ** 2)
    print(f'  平均功率降低: {power_reduction * 100:.2f}%')
    
    # 保存结果到本地目录
    print('\n保存结果...')
    os.makedirs(RESULTS_DIR, exist_ok=True)
    
    scipy.io.savemat(PREDICTION_SAVE_PATH, {
        'cleaned_eeg': predictions,
        'extracted_eog': artifacts,
        'original': test_x,
        'time_per_sample': time_per_sample,
        'consistency_error': consistency_error,
        'power_reduction': power_reduction,
        'sampling_rate': SAMPLING_RATE,
        'window_size': WINDOW_SIZE,
    })
    print(f'  ✓ 本地结果已保存: {PREDICTION_SAVE_PATH}')
    
    # 同时保存到总结果目录
    os.makedirs(FINAL_RESULTS_DIR, exist_ok=True)
    shutil.copy(PREDICTION_SAVE_PATH, FINAL_PREDICTION_PATH)
    print(f'  ✓ 总结果已保存: {FINAL_PREDICTION_PATH}')
    print(f'    - 去噪 EEG 形状: {predictions.shape}')
    print(f'    - 提取 EOG 形状: {artifacts.shape}')
    
    print('\n' + '='*80)
    print('测试完成！')
    print('='*80)


if __name__ == '__main__':
    main()

File: 复现的方法/ACMD/真实数据集测试/test_real_data.m
%% ACMD 真实数据集测试脚本
% 使用ACMD方法对真实数据集进行EOG伪影去除
% 输出统一格式的.mat文件

clear; clc;

%% 配置
% 数据路径
REAL_DATA_PATH = 'D:\Pycharm_Projects\EOG Remove\真实数据集\eog_dataset.mat';
DATA_KEY = 'eog_dataset';

% 采样率
SAMPLING_RATE = 250; % Hz

% 结果保存路径
RESULTS_DIR = 'results';
PREDICTION_SAVE_PATH = fullfile(RESULTS_DIR, 'ACMD_real_data_predictions.mat');

% 总结果保存路径
FINAL_RESULTS_DIR = fullfile('..', '..', '..', 'results');
FINAL_PREDICTION_PATH = fullfile(FINAL_RESULTS_DIR, 'ACMD_real_data_predictions.mat');

% 添加必要的路径
addpath(fullfile('..', '..'));  % 添加"复现的方法"目录以访问loadRealDatasetSplit

%% 加载真实数据集
fprintf('================================================================================\n');
fprintf('ACMD 真实数据集测试\n');
fprintf('================================================================================\n\n');

% 使用统一的数据划分函数（只加载测试集）
data = loadRealDatasetSplit(REAL_DATA_PATH, DATA_KEY);

%% 处理所有样本
n_samples = size(data, 1);
n_timepoints = size(data, 2);

fprintf('\n开始处理 %d 个样本...\n', n_samples);

% 初始化结果矩阵
cleaned_eeg = zeros(size(data));

tic; % 开始计时

for i = 1:n_samples
    % 获取当前样本
    signal = data(i, :);
    
    % 使用ACMD去除EOG伪影
    try
        % 调用ACMD去除函数
        cleaned_signal = oa_remove_acmd(signal, SAMPLING_RATE);
        cleaned_eeg(i, :) = cleaned_signal;
    catch ME
        fprintf('  ⚠️ 警告: 样本 %d 处理失败: %s\n', i, ME.message);
        % 如果处理失败，保留原始信号
        cleaned_eeg(i, :) = signal;
    end
    
    % 进度显示（每10个样本）
    if mod(i, 10) == 0
        elapsed = toc;
        avg_time = elapsed / i;
        remaining = avg_time * (n_samples - i);
        fprintf('  已处理 %d/%d 样本 (平均 %.3f秒/样本, 预计剩余 %.1f分钟)\n', ...
                i, n_samples, avg_time, remaining/60);
    end
end

total_time = toc;
time_per_sample = total_time / n_samples;

fprintf('\n处理完成!\n');
fprintf('  总耗时: %.1f秒\n', total_time);
fprintf('  单样本平均时间: %.3f秒\n', time_per_sample);

%% 计算伪影和统计信息
fprintf('\n计算统计信息...\n');

% 提取的EOG伪影
extracted_eog = data - cleaned_eeg;

% 验证解耦一致性
reconstructed = cleaned_eeg + extracted_eog;
consistency_error = mean((reconstructed - data).^2, 'all');
fprintf('  重建一致性 MSE: %.6f\n', consistency_error);

% 统计信息
original_std = std(data(:));
cleaned_std = std(cleaned_eeg(:));
artifact_std = std(extracted_eog(:));

fprintf('  原始信号标准差: %.4f\n', original_std);
fprintf('  去噪信号标准差: %.4f\n', cleaned_std);
fprintf('  伪影标准差: %.4f\n', artifact_std);

% 功率降低
power_reduction = (mean(data(:).^2) - mean(cleaned_eeg(:).^2)) / mean(data(:).^2);
fprintf('  平均功率降低: %.2f%%\n', power_reduction * 100);

%% 保存结果
fprintf('\n保存结果...\n');

% 创建结果目录
if ~exist(RESULTS_DIR, 'dir')
    mkdir(RESULTS_DIR);
end

% 保存到本地目录
original = data; % 重命名以匹配输出格式
save(PREDICTION_SAVE_PATH, ...
     'cleaned_eeg', 'extracted_eog', 'original', 'time_per_sample', ...
     'consistency_error', 'power_reduction', 'SAMPLING_RATE', 'n_timepoints', ...
     '-v7.3');

fprintf('  ✓ 本地结果已保存: %s\n', PREDICTION_SAVE_PATH);

% 保存到总结果目录
if ~exist(FINAL_RESULTS_DIR, 'dir')
    mkdir(FINAL_RESULTS_DIR);
end

% 重命名变量以匹配Python版本
original = data;
sampling_rate = SAMPLING_RATE;
window_size = n_timepoints;

save(FINAL_PREDICTION_PATH, ...
     'cleaned_eeg', 'extracted_eog', 'original', 'time_per_sample', ...
     'consistency_error', 'power_reduction', 'sampling_rate', 'window_size', ...
     '-v7.3');

fprintf('  ✓ 总结果已保存: %s\n', FINAL_PREDICTION_PATH);
fprintf('    - 去噪 EEG 形状: [%d, %d]\n', size(cleaned_eeg, 1), size(cleaned_eeg, 2));
fprintf('    - 提取 EOG 形状: [%d, %d]\n', size(extracted_eog, 1), size(extracted_eog, 2));

fprintf('\n================================================================================\n');
fprintf('测试完成！\n');
fprintf('================================================================================\n');
